CRM/Admin/Form/Setting.php

```diff
@@ -116,7 +117,11 @@ public function buildQuickForm() {
           );
         }
         elseif ($add == 'addSelect') {
-          $this->addElement('select', $setting, ts($props['title']), $options['values'], CRM_Utils_Array::value('html_attributes', $props));
+          $element = $this->addElement('select', $setting, ts($props['title']), $options['values'], CRM_Utils_Array::value('html_attributes', $props));
+          if ($setting == 'environment' && defined('CIVICRM_ENVIRONMENT')) {
+            $element->freeze();
+            CRM_Core_Session::setStatus(ts('The environment settings have been disabled because it has been overridden in the settings file.'), ts('Environment settings'), 'info');
+          }
```

Comment:
I really think this code belongs in the`CRM_Admin_Form_Setting_Debugging` class. After it calls `parent::buildQuickform()` it can do something like:

    if (defined('CIVICRM_ENVIRONMENT')) {
      $element = $this->getElement('environment');
      $element->freeze();
      CRM_Core_Session::setStatus(ts('The environment settings have been disabled because it has been overridden in the settings file.'), ts('Environment settings'), 'info');
    }